pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -29,188 +33,339 @@ type DockerRegistryServiceControllerOptions struct {
-		utilruntime.HandleError(err)
+	for {
+		if workFn() {
+			return
+		}
 	}
 }
 
-// serviceDeleted reacts to the docker-registry deletion by updating all the generated dockercfg secrets
-func (e *DockerRegistryServiceController) serviceDeleted(obj interface{}) {
-	service, ok := obj.(*api.Service)
-	if !ok {
-		return
+// getDockerRegistryLocations returns the dns form and the ip form of the secret
+func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string {
+	key, err := controller.KeyFunc(&kapi.Service{ObjectMeta: kapi.ObjectMeta{Name: e.serviceName, Namespace: e.serviceNamespace}})
+	if err != nil {
+		return []string{}
+	}
+
+	obj, exists, err := e.serviceCache.GetByKey(key)
+	if err != nil {
+		return []string{}
 	}
-	if service.Name != e.registryServiceName {
-		return
+	if !exists {
+		return []string{}
 	}
+	service := obj.(*kapi.Service)
 
-	if err := e.handleLocationChange(e.defaultDockerURL); err != nil {
-		utilruntime.HandleError(err)
+	hasPortalIP := (len(service.Spec.ClusterIP) > 0) && (net.ParseIP(service.Spec.ClusterIP) != nil)
+	if hasPortalIP && len(service.Spec.Ports) > 0 {
+		return []string{
+			net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+			net.JoinHostPort(fmt.Sprintf("%s.%s.svc", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+		}
 	}
+
+	return []string{}
 }
 
 // handleLocationChange goes through all service account dockercfg secrets and updates them to point at a new docker-registry location
-func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
-	e.dockercfgController.SetDockerURL(serviceLocation)
-
-	dockercfgSecrets, err := e.listDockercfgSecrets()
-	if err != nil {
-		return err
+func (e *DockerRegistryServiceController) syncRegistryLocationChange(key string) error {
+	newDockerRegistryLocations := sets.NewString(e.getDockerRegistryLocations()...)
+	if e.getRegistryURLs().Equal(newDockerRegistryLocations) {
+		glog.V(4).Infof("No effective update: %v", newDockerRegistryLocations)
+		return nil
 	}
 
-	for _, dockercfgSecret := range dockercfgSecrets {
-		dockercfg := &credentialprovider.DockerConfig{}
-		if err := json.Unmarshal(dockercfgSecret.Data[api.DockerConfigKey], dockercfg); err != nil {
-			utilruntime.HandleError(err)
-			continue
-		}
+	// make sure that new dockercfg secrets get the correct locations
+	e.dockercfgController.SetDockerURLs(newDockerRegistryLocations.List()...)
+	e.setRegistryURLs(newDockerRegistryLocations.List()...)
 
-		dockercfgMap := map[string]credentialprovider.DockerConfigEntry(*dockercfg)
-		keys := sets.StringKeySet(dockercfgMap)
-		if len(keys) != 1 {
-			utilruntime.HandleError(err)
+	// we've changed the docker registry URL.  Add items to the work queue for all known secrets
+	// new secrets will already get the updated value.
+	for _, obj := range e.secretCache.List() {
```

Comment:
Naive question: are all dockercfg secrets in the cluster related to the docker registry?